cache decorator

JohnDoePlusPlus opened the pull request.

Commit:
cache

dimadeveatii commented on package.json:
```diff
@@ -39,6 +39,7 @@
     "@types/chai": "4.1.7",
     "@types/chai-as-promised": "7.1.0",
     "@types/mocha": "5.2.5",
+    "@types/object-hash": "^1.2.0",
```
Use same conventions as other version specifiers.

dimadeveatii commented on test/utils.ts:
```diff
@@ -11,3 +11,9 @@ export function repeat(func: () => void, count: number) {
   return results;
 }
+
+export async function executionTime(func: () => Promise<any>): Promise<number> {
```
Avoid `any`. Either use generics, or `never` type if you don't use the return type.

dimadeveatii commented on lib/cache/storages/MemoryStorage.ts:
```diff
@@ -0,0 +1,32 @@
+  public set<V>(key: string, value: V): Storage {
+    if (typeof this.limit === 'number' && this.storage.size >= this.limit) {
+      return this;
+    }
```
I'd expect that when limit is exceeded, then the value is still cached, but the oldest entry is removed.  That means that when a limit is specified, you also use a bounded list to track which values must be discarded.

dimadeveatii commented on lib/cache/storages/Storage.ts:
```diff
@@ -0,0 +1,6 @@
+  get<V>(key: string): V;
+  has(key: string): boolean;
+  delete(key: string): Storage;
+}
```
The interface is synchronous, limiting the implementations to only in-memory strategies. The interface definition will not fit an implementation for a Redis or MemCache caches.

dimadeveatii commented on lib/cache/hash/hash.ts:
```diff
@@ -0,0 +1,9 @@
+
+export class HashService {
+
+  public hash(value: any): string {
```
To avoid constructions of the form `hash.hash()` I'd suggest that we rename the method to `.calculate()`

dimadeveatii replied:
And the `HashService` seems to be a strange in the `cache` domain. I'd suggest moving it to some kind of utils outside of `cache`.

dimadeveatii commented on lib/cache/expirations/Expiration.ts:
```diff
@@ -0,0 +1,3 @@
+export interface Expiration {
+  add(key: string, clear: () => any): void;
```
Why return `any`? is it used further?
At the same time, would be nice if clear would be called with the `key`.

dimadeveatii commented on lib/cache/expirations/AbsoluteExpiration.ts:
```diff
@@ -0,0 +1,27 @@
+
+  private clear(key: string, clear: () => any): () => void {
+    return () => {
+      clear();
```
`clear` is an external function that might fail, in which case we're introducing a memory leak. I'd suggest that we first delete the key, and then invoke `clear()`.

dimadeveatii commented on lib/cache/expirations/SlidingExpiration.ts:
```diff
@@ -0,0 +1,35 @@
+  private addKey(key: string, clear: () => any): void {
+    const timeoutId = setTimeout(
+      () => {
+        clear();
```
Same as in other comment, call `clear()` at the end.

dimadeveatii commented on lib/cache/expirations/SlidingExpiration.ts:
```diff
@@ -0,0 +1,35 @@
+      this.timeout,
+    );
+
+    this.expirations.set(key, [timeoutId as any, clear]);
```
Do we need to keep the `clear` functions?

Commit:
cahce fixes

Commit:
cache decorator improve tests

dimadeveatii commented on examples/cache.ts:
```diff
@@ -0,0 +1,18 @@
+
+service.asyncMethod(3).then(res => console.log(res)); // first call => no cache
+service.asyncMethod(3).then(res => console.log(res)); // second call => from cache
+service.asyncMethod(4).then(res => console.log(res)); // first call => no cache
```
`first` -> `third`. Not clear, why not cache here?

dimadeveatii commented on lib/cache/caches/Cache.ts:
```diff
@@ -0,0 +1,6 @@
+
+export interface Cache<K> {
+  set<V>(key: K, value: V, instance: ClassType): Promise<void>;
+  get<V>(key: K, instance: ClassType): Promise<V>;
```
What is `instance` used for?

dimadeveatii commented on lib/cache/index.ts:
```diff
@@ -0,0 +1,68 @@
+
+export { CacheOptions };
+
+interface CacheOptionsAndTimeout extends CacheOptions {
```
I'd suggest better to rename it to `TimeoutCacheOptions`. At the same time `CacheOptions` is a type, why making this one an interface and not type (for consistency)?

dimadeveatii commented on lib/cache/index.ts:
```diff
@@ -0,0 +1,68 @@
+interface Parameters {
+  timeout: number;
+  options: CacheOptions;
+}
```
I'd just remove this one and not use it at all. For `parseParameters` function use the implicit return type.

dimadeveatii commented on lib/cache/storages/Storage.ts:
```diff
@@ -0,0 +1,5 @@
+export interface Storage {
+  set<V>(key: string, value: V): Promise<void>;
```
Would be nice to make the api chainable and return `Promise<this>` instead of `Promise<void>`. Same for `delete` method.

dimadeveatii commented on lib/cache/storages/MemoryStorage.ts:
```diff
@@ -0,0 +1,50 @@
+
+    if (this.hasLimit) {
+      this.keys = this.keys.filter(value => value !== key);
+    }
```
Suboptimal implementation. The delete operation results in `O(N)` complexity and a new memory allocation for the array.

dimadeveatii commented on lib/cache/storages/MemoryStorage.ts:
```diff
@@ -0,0 +1,50 @@
+export class MemoryStorage implements Storage {
+
+  private readonly storage = new Map<string, any>();
+  private keys: string[];
```
Array type is not the best choice here. Use a more optimal data type.

dimadeveatii commented on lib/cache/storages/MemoryStorage.ts:
```diff
@@ -0,0 +1,50 @@
+    return Promise.resolve(this.storage.get(key));
+  }
+
+  public delete(key: string = this.keys[0]): Promise<void> {
```
Using a default value for `key` here is not intuitive. It is recommended to keep the same signature as in the original interface.

dimadeveatii commented on lib/utils/hash/hash.ts:
```diff
@@ -0,0 +1,9 @@
+import * as hash from 'object-hash';
```
I'd suggest to rename `hash.ts` to `index.ts` or just implement directly as `lib/utils/hash.ts`

dimadeveatii commented on lib/cache/expirations/AbsoluteExpiration.ts:
```diff
@@ -0,0 +1,27 @@
+    private readonly timeout: number,
+  ) { }
+
+  public add(key: string, clear: (key: string) => unknown): void {
```
Maybe `never` is more appropriate than `unknown`? And for `clear` maybe `clearCallback` is more suggestive.

dimadeveatii commented on lib/cache/expirations/AbsoluteExpiration.ts:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    this.expirations.add(key);
+    setTimeout(this.clear(key, clear), this.timeout);
```
Simplify code, don't return a function in `clear` and use directly: `setTimeout(() => this.clear(key, clear), this.timeout);`

dimadeveatii commented on lib/cache/caches/ClassCache.ts:
```diff
@@ -0,0 +1,32 @@
+  public async get<V>(key: K): Promise<V> {
+    const keyHash = this.hash.calculate(key);
+
+    this.expiration.add(keyHash, key => this.delete(key));
```
At this point the key might not be cached any more, still you add it to the expiration. I'd suggest first to check if the key is in the storage.

dimadeveatii commented on lib/cache/factories/storeFactory.ts:
```diff
@@ -0,0 +1,19 @@
+import { CacheOptions } from '..';
```
`store` filename is not inline with `storage` used in other files.

dimadeveatii commented on lib/cache/factories/cacheFactory.ts:
```diff
@@ -0,0 +1,44 @@
+import { CacheOptions } from '..';
```
I consider that all factory implementation shouldn't be separated in a specific folder, but rather be part of their domain like `cache/factories/cacheFactory.ts` -> `cache/caches/factory.ts`

Commit:
cache
move scope logic from cache to new class cache manager

dimadeveatii commented on lib/cache/cacheManager/CacheManager.ts:
```diff
@@ -0,0 +1,6 @@
+import { ClassType } from '../../interfaces/class';
+import { Cache } from '../caches/Cache';
+
+export interface CacheManager<K = any> {
```
It sounds more like a `CacheProvider` rather than a manager.

dimadeveatii commented on lib/cache/cacheManager/InstanceCacheManager.ts:
```diff
@@ -0,0 +1,26 @@
+
+export class InstanceCacheManager<K = any> implements CacheManager<K> {
+
+  private cacheByInstances = new WeakMap<ClassType, Cache<K>>();
```
Maybe `instanceCaches` is better?

dimadeveatii commented on lib/cache/cacheManager/InstanceCacheManager.ts:
```diff
@@ -0,0 +1,26 @@
+  ) { }
+
+  public get(instance: ClassType): Cache<K> {
+    const shouldCreateCache = !this.cacheByInstances.has(instance);
```
`hasCache` or `isCached`, always prefer short names easy to read and understand.

dimadeveatii commented on lib/cache/cacheManager/ClassCacheManager.ts:
```diff
@@ -0,0 +1,23 @@
+
+export class ClassCacheManager<K = any> implements CacheManager<K> {
+
+  private cacheInstance: Cache<K> = null;
```
the suffix `Instance` here is useless [Clean Code].

dimadeveatii commented on lib/cache/cacheManager/CacheManager.ts:
```diff
@@ -0,0 +1,6 @@
+import { Cache } from '../caches/Cache';
+
+export interface CacheManager<K = any> {
+  get(instance: ClassType): Cache<K>;
```
I think `instance` should be optional here.

dimadeveatii commented on lib/cache/cacheManager/factory.ts:
```diff
@@ -0,0 +1,27 @@
+  <K>(timeout: number, options: CacheOptions) => CacheManager<K>
+> = new Map<'class' | 'instance', <K>(timeout: number, options: CacheOptions) => CacheManager<K>>()
+  .set('class', (timeout, options) => new ClassCacheManager<any>(timeout, options))
+  .set('instance', (timeout, options) => new InstanceCacheManager<any>(timeout, options));
```
To be honest this is hard to read. Do you need this map at all? Maybe plain vanilla `if/else` or `switch` ?

dimadeveatii commented on lib/cache/cacheManager/ClassCacheManager.ts:
```diff
@@ -0,0 +1,23 @@
+
+  public get(): Cache<K> {
+    if (!this.cacheInstance) {
+      this.cacheInstance = cacheFactory(this.timeout, this.options);
```
Because `cacheFactory` is not injected, it makes it hard to unit test this class.

dimadeveatii commented on lib/cache/caches/factory.ts:
```diff
@@ -0,0 +1,13 @@
+import { storageFactory } from '../storages/factory';
+import { Cache } from './Cache';
+
+export function cacheFactory<K = any>(timeout: number, options: CacheOptions): Cache<K> {
```
How would you write unit tests for this?

dimadeveatii commented on lib/cache/expirations/factory.ts:
```diff
@@ -0,0 +1,21 @@
+const expirationFactories: ReadonlyMap<'absolute' | 'sliding', (timeout: number) => Expiration> =
+  new Map<'absolute' | 'sliding', (timeout: number) => Expiration>()
+    .set('absolute', timeout => new AbsoluteExpiration(timeout))
+    .set('sliding', timeout => new SlidingExpiration(timeout));
```
I'd prefer a switch statement instead.

dimadeveatii commented on lib/cache/index.ts:
```diff
@@ -0,0 +1,65 @@
+
+    descriptor.value = async function (...args: any[]) {
+      const cacheService = cacheManager.get(this);
+      const wasCached = await cacheService.has(args);
```
Why `wasCached`? We need to check if still `isCached` 😏 ?

dimadeveatii commented on lib/cache/index.ts:
```diff
@@ -0,0 +1,65 @@
+        cacheService.set(args, value);
+        return value;
+      } catch (error) {
+        return Promise.reject(error);
```
Catch seems to be useless here, you just propagate the error.

dimadeveatii commented on lib/cache/storages/MemoryStorage.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  public async delete(key: string): Promise<this> {
+    this.storage.delete(key);
```
Shouldn't you also delete the `key` from the `keysStorage`?

JohnDoePlusPlus replied:
I change implementation for this logic, hope now is better

dimadeveatii commented on lib/cache/storages/MemoryStorage.ts:
```diff
@@ -0,0 +1,55 @@
+
+    if (this.hasLimit) {
+      this.keysStorage.delete(key);
+    }
```
When I explicitly delete a key, do I need to check the limit?

dimadeveatii reviewed (changes requested):
Please cover with unit tests each internal file.

Commit:
cache
replace factory functions to class
rename CacheManager to CacheProvider
remove keysStorageIn MemoryStorage (now is using keys from map)

Commit:
cache
improve code in factories and in SlidingExpiration
add unit tests

dimadeveatii commented on lib/cache/cacheProvider/InstanceCacheProvider.ts:
```diff
@@ -0,0 +1,22 @@
+
+  public get(instance: ClassType): Cache<K> {
+    const hasCache = !this.instanceCaches.has(instance);
+    if (hasCache) {
```
The variable name lies 😃 .

JohnDoePlusPlus replied:
everybody lies :)

dimadeveatii commented on lib/cache/index.ts:
```diff
@@ -0,0 +1,69 @@
+  const expirationFactory = new ExpirationFactory(timeout, options.expiration);
+  const storageFactory = new StorageFactory(options.size, options.storage);
+  const cacheFactory = new CacheFactory(hashService, expirationFactory, storageFactory);
+  const cacheProvider = new CacheProviderFactory(options.scope, cacheFactory).create();
```
In the end only the `cacheProvider` is used. To make the decorator code cleaner, please move above lines in a separate function.

dimadeveatii commented on test/cache/cacheProvider/ClassCacheProvider.spec.ts:
```diff
@@ -0,0 +1,46 @@
+
+    it('should create', () => expect(service).to.be.instanceOf(ClassCacheProvider));
+
+    it('should init cache property to null', () => expect(service['cache']).to.be.null);
```
Do we need this test?

dimadeveatii commented on test/cache/cacheProvider/ClassCacheProvider.spec.ts:
```diff
@@ -0,0 +1,46 @@
+      const cacheInstance = {} as any;
+      cacheFactoryStub.create.returns(cacheInstance);
+
+      expect(service.get()).to.be.equals(cacheInstance);
```
`to.be.equals()` -> `to.equal()`. At least for me sounds easier to read.

dimadeveatii commented on test/cache/cacheProvider/ClassCacheProvider.spec.ts:
```diff
@@ -0,0 +1,46 @@
+
+      expect(service.get()).to.be.equals(cacheInstance);
+      expect(cacheFactoryStub.create.calledOnce).to.be.true;
+      expect(service['cache']).to.be.equals(cacheInstance);
```
Do not try to test internal state.

dimadeveatii replied:
Try to make one assertion per test. Check the unit testing FIRST principle.

dimadeveatii reviewed (changes requested):
Please review unit tests, make sure you don't test internal state, only the public interface. Each unit test should have one assertion.

Commit:
cache
improve unit tests

Commit:
cache
improve unit tests

Commit:
Merge branch 'story/cache' of github.com:JohnDoePlusPlus/micro-decorators into story/cache

dimadeveatii commented on lib/cache/storages/MemoryStorage.ts:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  private checkSize(): void {
+    if (typeof this.limit === 'number' && this.storage.size >= this.limit) {
```
To avoid the check for type of limit, I'd suggest that we default it to a very large number, like `Number.POSITIVE_INFINITY`

dimadeveatii commented on test/uitls/hash.spec.ts:
```diff
@@ -0,0 +1,57 @@
+      const secondObject = { b: 2 };
+
+      expect(service.calculate(firstObject)).not.equals(service.calculate(secondObject));
+    });
```
Please add tests for array of objects, with same and with different keys (2 tests are enough).

dimadeveatii commented on test/utils.ts:
```diff
@@ -11,3 +11,9 @@ export function repeat(func: () => void, count: number) {
   return results;
 }
+
+export async function executionTime(func: () => Promise<unknown>): Promise<number> {
```
Is this used and do we need it for unit testing?

JohnDoePlusPlus replied:
It is used to test cache decorator.

dimadeveatii commented on lib/cache/storages/factory.ts:
```diff
@@ -0,0 +1,22 @@
+    private readonly storage: 'memory',
+  ) { }
+
+  public create(): Storage {
```
I'd recommend that the `storage` type is passed as an argument to the `create` function instead of in constructor.

dimadeveatii commented on lib/cache/expirations/factory.ts:
```diff
@@ -0,0 +1,26 @@
+    private readonly expiration: 'absolute' | 'sliding',
+  ) { }
+
+  public create(): Expiration {
```
Same as for storage, `expiration` is recommended to be part of the `create` method.

dimadeveatii commented on test/cache/caches/factory.spec.ts:
```diff
@@ -0,0 +1,48 @@
+    expirationFactoryStub = sinon.createStubInstance(ExpirationFactory);
+    storageFactoryStub = sinon.createStubInstance(StorageFactory);
+
+    service = new CacheFactory(undefined, expirationFactoryStub as any, storageFactoryStub as any);
```
Avoid use of `any`. Pass also a stub for hash service, do not use undefined.

dimadeveatii commented on test/cache/expirations/AbsoluteExpiration.spec.ts:
```diff
@@ -0,0 +1,68 @@
+
+      it('should not update expiration', async () => {
+        const spy = sinon.spy();
+        service['expirations'].add('key');
```
Avoid using internals.

dimadeveatii commented on test/cache/expirations/AbsoluteExpiration.spec.ts:
```diff
@@ -0,0 +1,68 @@
+        await delay(timeout / 2);
+
+        expect(firstSpy.calledOnce).to.be.true;
+      });
```
Add tests that actually verify that cached value is returned before expiration, and that `undefined` is returned after expiration.

dimadeveatii commented on test/cache/expirations/SlidingExpiration.spec.ts:
```diff
@@ -0,0 +1,64 @@
+
+    describe('existing key', () => {
+
+      beforeEach(() => service['expirations'].set('key', setTimeout(() => { }, timeout) as any));
```
Avoid using internals.

dimadeveatii commented on test/cache/expirations/SlidingExpiration.spec.ts:
```diff
@@ -0,0 +1,64 @@
+        await delay(timeout);
+
+        expect(spy.calledOnce).to.be.true;
+      });
```
Add tests that actually assert that cached value is returned when using *sliding access* and that `undefined` after a timeout.

Commit:
cache
change factories, not create is with parameters
improve unit tests

dimadeveatii commented:
We should analyze the possibility to use the popular library [lru-cache](https://www.npmjs.com/package/lru-cache).
